### PR TITLE
NEAR Indexer for Wallet

### PR DESCRIPTION
**Pull request by khorolets**
NEAR Wallet Indexer PoC 

> Syncing all the AddKey actions from successful receipts into a table (public_key -> receiver_id (account id), and we can have multiple public keys for a single account as well as the same public key can be [it is not recommended, but it is not restricted] used by several accounts)

@frol: Given that we can have multiple AddKey and DeleteKey actions throughout the history... [skipped]
We are going to design the Wallet Indexer to expose a single table with the following columns:
```
public key
account id
action (ADD | DELETE)
status (PENDING | SUCCESS | FAILED)
receipt id (we need it to connect Execution Outcome status which only gets known in the next block after we observe the receipt itself)
block height (this way you can get the latest state)
permission (FUNCTION_CALL | FULL_ACCESS)
```
To query the public key you will need to do the following query:
```sql
SELECT action
FROM access_keys
WHERE public_key = :public_key AND account_id = :account_id AND status = "SUCCESS"
ORDER BY block_height DESC
LIMIT 1
```
If the `action` is `"ADD"`, then the latest state is that the key exists and belongs to the account, otherwise, it used to be there, but not anymore.

---

Also before start NEAR Wallet Indexer handles AccessKey addition from genesis, so all the keys from genesis will be added to database.

`tools/indexer/wallet/.env` should be added with your credentials to PostgreSQL database.

Example:
```
DATABASE_URL=postgres://user:password@host/database_name
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,28 @@ Cargo.lock
 
 # These are backup files generated by rustfmt
 **/*.rs.bk
+
+# These are backup files generated by rustfmt
+**/*.rs.bk
```
Comment (frol) [inline .gitignore]: This is a duplicate

File: Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+bigdecimal = "0.1.0"
+diesel = { version = "1.4.4", features = ["postgres", "numeric", "serde_json"] }
+dotenv = "0.15.0"
+r2d2 = "0.8.8"
+serde_json = "1.0.55"
+tokio = { version = "0.2", features = ["full"] }
+tokio-diesel = "0.3.0"
+
+near-indexer = { git = "https://github.com/nearprotocol/nearcore" }
+# Required until https://github.com/nearprotocol/nearcore/pull/3000 merged
```
Comment (frol) [inline Cargo.toml]: Done

**Comment by khorolets** [comment]
@frol comment (from https://github.com/nearprotocol/nearcore/pull/2999)
> Let's call it NEAR_INDEXER_FOR_WALLET_DATABASE_URL, otherwise, we may collide with something completely different.

It seems to be not configurable from Diesel ORM side, it required this particular name in `.env`

File: src/db/access_keys.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+impl AccessKey {
+    pub fn from_receipt_view(
+        receipt: &near_indexer::near_primitives::views::ReceiptView,
+        block_height: u64,
+    ) -> Vec<Self> {
+        let mut access_keys: Vec<Self> = vec![];
+        if let near_indexer::near_primitives::views::ReceiptEnumView::Action { actions, .. } =
+        &receipt.receipt
+        {
```
Comment (frol) [inline src/db/access_keys.rs]: Is this rustfmt formatting? OMG it is weird! Can you give `cargo fmt` a try?

File: src/db/access_keys.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                        receipt_hash: receipt.receipt_id.to_string(),
+                        block_height: block_height.into(),
+                        permission: match access_key.permission {
+                            near_indexer::near_primitives::views::AccessKeyPermissionView::FullAccess => {
+                                PermissionEnum::FullAccess
+                            },
+                            near_indexer::near_primitives::views::AccessKeyPermissionView::FunctionCall { .. } => {
+                                PermissionEnum::FunctionCall
+                            },
+                        },
```
Comment (frol) [inline src/db/access_keys.rs]: You can implement `from AccessKeyPermissionView for PermissionEnum`, and then use `permission: access_key.permission.into()` here and in the `main.rs` (genesis handler).

File: src/main.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+async fn handle_genesis_public_keys(near_config: near_indexer::NearConfig) {
+    let pool = establish_connection();
+    let genesis_height = near_config.genesis.config.genesis_height.clone();
+    let access_keys = near_config.genesis.records.as_ref()
+        .iter()
+        .filter(|record| match record {
+        near_indexer::near_primitives::state_record::StateRecord::AccessKey { .. } => true,
+        _ => false,
+    })
```
Comment (frol) [inline src/main.rs]: This is redundant, just remove it.

```suggestion
```

File: src/main.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                    schema::access_keys::dsl::receipt_hash.eq_any(
+                        receipt_outcomes
+                            .iter()
+                            .filter_map(|outcome| match &outcome.outcome.status {
+                                near_indexer::near_primitives::views::ExecutionStatusView::Failure(
+                                    _,
+                                ) => Some(outcome.id.to_string()),
+                                _ => None,
+                            })
+                            .collect::<Vec<String>>(),
```
Comment (frol) [inline src/main.rs]: The nesting is too deep. Let's try to extract it to a separate variable above the match.

**Review by frol — APPROVED** [review]
I love it!

**Pull request merged**